ksqldb-engine/src/main/java/io/confluent/ksql/engine/InsertValuesExecutor.java

```diff
@@ -266,12 +274,52 @@ private void throwIfDisabled(final KsqlConfig config) {
           FormatFactory.fromName(dataSource.getKsqlTopic().getKeyFormat().getFormat()),
           topicName,
           true,
+          "write",
           e);
       LOG.error("Could not serialize key.", e);
       throw new KsqlException("Could not serialize key: " + keyValue, e);
     }
   }
 
+  /**
+   * Ensures that the key schema that we generate will be identical
+   * to the schema that is registered in schema registry, if it exists.
+   * Otherwise, it is possible that we will publish messages with a new
+   * schemaID, meaning that logically identical keys might be routed to
+   * different partitions.
+   */
+  private void ensureKeySchemasMatch(
+      final PersistenceSchema keySchema,
+      final DataSource dataSource,
+      final ServiceContext serviceContext
+  ) {
+    final KeyFormat keyFormat = dataSource.getKsqlTopic().getKeyFormat();
+    final Format format = FormatFactory.fromName(keyFormat.getFormat());
+    if (!format.supportsFeature(SerdeFeature.SCHEMA_INFERENCE)) {
+      return;
+    }
+
+    final ParsedSchema schema = format
+        .getSchemaTranslator(keyFormat.getFormatInfo().getProperties())
+        .toParsedSchema(keySchema);
+
+    try {
+      final SchemaMetadata latest = serviceContext.getSchemaRegistryClient()
+          .getLatestSchemaMetadata(
```

Comment:
How come we're not using `SchemaRegistryUtil#getLatestSchema()` here?

Reply:
umm, I wasn't aware that existed!

Reply:
also 🤦 seems like I created it...

Reply:
https://github.com/confluentinc/ksql/pull/6632